ios/BetterRailWidget/Live Activity/ActivityUtils.swift

```diff
@@ -0,0 +1,225 @@
+    return nil;
+  }
+}
+
+/// Get information about the current ride progress
+/// - Returns: A tuple made out of the (current stop station index, total stop station count)
+func rideProgress(route: Route, nextStationId: Int) -> (Int, Int) {
+  if let train = getTrainFromStationId(route: route, stationId: nextStationId) {
+    let totalStations = train.stopStations.count + 1
+    if let currentIndex = train.stopStations.firstIndex(where: { $0.stationId == nextStationId }) {
+      return (currentIndex, totalStations)
+    } else {
+      /// Check if next station id is the last stop
+      for index in 0 ..< route.trains.count {
+        if (nextStationId == route.trains[index].destinationStation) {
+          let stopStationsCount = route.trains[index].stopStations.count + 1
+          return (stopStationsCount - 1, stopStationsCount)
+        }
+      }
+      
+      print("Failed to find the last stop station in route")
+      return (0, 0);
+    }
+  } else {
+    print("Failed to find the last stop station in route")
+    return (0, 0);
+  }
+}
+
+/// Get the start date for the current status
+/// e.g. for `waitForTrain` we'll use the activity start date, and for `inTransit` we'd want to use the train departure time
+@available(iOS 16.1, *)
+func getStatusStartDate(context: ActivityViewContext<BetterRailActivityAttributes>) -> Date {
+  let status = context.state.status;
+  let route = context.attributes.route
+  let nextStationId = context.state.nextStationId
+  let train = getTrainFromStationId(route: route, stationId: nextStationId)!
+  let delay = context.state.delay
+  let departureDate = isoDateStringToDate(train.departureTime).addDelay(delay)
+  
+  if (status == .waitForTrain) {
+    return context.attributes.activityStartDate
+  } else if (status == .inExchange) {
+    let previousTrain = getPreviousTrainFromStationId(route: route, stationId: nextStationId)
+    // since we're in exchange, use the arrival time of the previous train as the progress start time
+    return isoDateStringToDate(previousTrain?.arrivalTime ?? "").addDelay(delay)
+  } else {
+    return departureDate
+  }
+}
+
+@available(iOS 16.1, *)
+func getStatusEndDate(context: ActivityViewContext<BetterRailActivityAttributes>) -> Date {
+  let status = context.state.status
+  let route = context.attributes.route
+  let nextStationId = context.state.nextStationId
+  let delay = context.state.delay
+  let train = getTrainFromStationId(route: route, stationId: nextStationId)!
+  
+  let departureDate = isoDateStringToDate(train.departureTime).addDelay(delay)
+  let arrivalDate = isoDateStringToDate(train.arrivalTime).addDelay(delay)
+
+  if (status == .waitForTrain || status == .inExchange) {
+    return departureDate
+  } else {
+    return arrivalDate
+  }
+}
+
+func getMinutesLeft(targetDate: Date) -> Int {
+  let now = Date()
+  let timeInterval = targetDate.timeIntervalSince(now)
+  let minutes = Int(round(timeInterval / 60))
+
+  if minutes <= 0 {
+    return 1
```

Comment:
I think if such a scenario occurs, we should show "Now".

Reply:
I prefer to avoid displaying "Now", and instead to hide the counter once the status changes to `arrived` - and to keep displaying "1 min" until then. wdyt?